libdnf/base/goal.cpp

```diff
@@ -596,10 +604,136 @@ void Goal::Impl::add_rpms_to_goal(base::Transaction & transaction) {
                     settings.resolve_best(cfg_main),
                     settings.resolve_clean_requirements_on_remove());
                 break;
+            case GoalAction::REINSTALL: {
+                solv::IdQueue ids_nevra_installed;
+                for (auto id : ids) {
+                    rpm::PackageQuery query(installed);
+                    query.filter_nevra({pool.get_nevra(id)});
+                    if (query.empty()) {
+                        // Report when package with the same NEVRA is not installed
+                        transaction.p_impl->add_resolve_log(
+                            GoalAction::REINSTALL,
+                            GoalProblem::NOT_INSTALLED,
+                            settings,
+                            {pool.get_nevra(id)},
+                            {},
+                            false);
+                    } else {
+                        // Only installed packages can be reinstalled
+                        ids_nevra_installed.push_back(id);
+                    }
+                }
+                rpm_goal.add_install(
+                    ids_nevra_installed,
+                    settings.resolve_strict(cfg_main),
+                    settings.resolve_best(cfg_main),
+                    settings.resolve_clean_requirements_on_remove());
+            } break;
             case GoalAction::UPGRADE: {
+                solv::IdQueue ids_upgrades;
+                for (auto id : ids) {
+                    rpm::PackageQuery query(installed);
+                    query.filter_name({pool.get_name(id)});
+                    if (query.empty()) {
+                        // Report when package with the same name is not installed
+                        transaction.p_impl->add_resolve_log(
+                            GoalAction::UPGRADE,
+                            GoalProblem::NOT_INSTALLED,
+                            settings,
+                            {pool.get_nevra(id)},
+                            {},
+                            false);
+                    } else {
+                        query.filter_arch({pool.get_arch(id)});
+                        if (query.empty()) {
+                            // Report when package with the same name is installed for a different architecture
+                            transaction.p_impl->add_resolve_log(
+                                GoalAction::UPGRADE,
+                                GoalProblem::NOT_INSTALLED_FOR_ARCHITECTURE,
+                                settings,
+                                {pool.get_nevra(id)},
+                                {},
+                                false);
+                        } else {
+                            rpm::PackageQuery query_id(base, rpm::PackageQuery::ExcludeFlags::IGNORE_EXCLUDES, true);
```

Comment:
I am not sure whether we should report it at all - version EQ or higher already installed. Anyway when version of the package is the same with version of installed package I would prefer to make a decision by solver. There are possibilities when solver decide that the package should be replaced.

Reply:
It is reported now (as info), but the package is passed to the solver anyway.